trackintel/preprocessing/staypoints.py

```diff
@@ -169,3 +178,28 @@ def generate_locations(
     ret_loc["user_id"] = ret_loc["user_id"].astype(ret_stps["user_id"].dtype)
 
     return ret_stps, ret_loc
+
+
+def _generate_locations_per_user(df, location_id_counter, distance_metric, db, geo_col):
+    """function called after groupby: should only contain records of one user; see generate_locations() function for parameter meaning."""
+
+    user_staypoints = df
+
+    # ensuring unique labels: we assume that every pandas group must have a unique index for the first element of the group
+    location_id_counter += user_staypoints.index[0]
+
+    if distance_metric == "haversine":
+        # the input is converted to list of (lat, lon) tuples in radians unit
+        p = np.array([[radians(q.y), radians(q.x)] for q in (user_staypoints[geo_col])])
+    else:
+        p = np.array([[q.x, q.y] for q in (user_staypoints[geo_col])])
+    labels = db.fit_predict(p)
+
+    # enforce unique lables across all users without changing noise labels
+    labels[labels != -1] = labels[labels != -1] + location_id_counter
+
+    # add staypoint - location matching to original staypoints
+    user_staypoints.loc[user_staypoints.index, "location_id"] = labels
```

Comment:
`user_staypoints.loc[user_staypoints.index, "location_id"]` is basically `user_staypoints["location_id"]`?

Reply:
yes, modified